### PR TITLE
Polymorphism in Java

### LINKED ISSUE #4746: Polymorphism in Java
### Introduction
Polymorphism is the ability of an object to have more than one form. In other words, polymorphism is a feature in Java that allows a single task to be performed in different ways.
An example of polymorphism is the human body. The human body has organs like heart to pump blood, stomach for food and lungs to breathe.
In this article, we will break down polymorphism and give a detailed explanation with clear examples.
We will take a look at how it is implemented in java and how it relates to inheritance.
Furthermore, we'll discuss different types of polymorphism and how they differ from one another. Run time polymorphism will also be discussed.
Lastly, we'll use a code snippet of a real life example to illustrate the point of polymorphism.

### Key Takeaways
By the end the article, the following should have been covered:
1. Introduction to Polymorphism.
2. Examples of polymorphism.
3. Implementation of Polymorphism in java.
4. Types of Polymorphism.
5. Run time Polymorphism.
6. Real Life example of Polymorphism.

#### Article quality
The article proposal will provide an in-depth  explanation on the topic. There will be use of code snippets flow charts where applicable to help the user understand. The language used will be clear easy to follow through. The content will help a beginner to fully understand what polymorphism entails.

#### Reference
N/A

### PR DESCRIPTION
**Pull request by Kimosop-Wakanda**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #4746

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by mohamedgh16** [comment]
@ahmadmardeni1 I would like to review this article.

**Comment by ahmadmardeni1** [comment]
Go ahead. @mohamedgh16

**Comment by ahmadmardeni1** [comment]
Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept.
Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content.
Be sure to see our resources page to see more info on plagiarism and what is considered as such.

[Polymorphism in Java#5864](https://github.com/section-engineering-education/engineering-education/files/7799056/New.Scan.12_22.PM.Copyleaks.Report.pdf)

And I think this article is light on substance. What do you think? @mohamedgh16

**Comment by Kimosop-Wakanda** [comment]
@ahmadmardeni1 
I'll add more content and check on the plagiarism right away
I'll ping you when am  done
cc: @mohamedgh16

**Comment by Kayere** [comment]
Plagiarism check done ✅

File: content/articles/polymorphism-in-java/index.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+        this.legs = this.legs + legs;
+    }
+}
+```
+
+Declaring of polymorphic parameters can lead to `variable hiding`.
+
+**Variable hiding** occurs where the declaration of a local variable overrides the global variable with the same name.
+
+In the above example, we have declared the legs globally as a string and then declared them locally as an integer. Normally variable hiding would have occurred. But to solve for that, we have used a global reference `this` to point to the global variables within the local context.
```
Comment (mohamedgh16) [inline content/articles/polymorphism-in-java/index.md:208]: Normally,

File: content/articles/polymorphism-in-java/index.md [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
```
Comment (mohamedgh16) [inline content/articles/polymorphism-in-java/index.md:209]: Your article is missing a conclusion, add it after the summary. You may also add a further reading.

**Review by mohamedgh16 — COMMENTED** [review]
Great article @Kimosop-Wakanda. Please finish the changes so we can move to the tech review.

File: content/articles/polymorphism-in-java/index.md (hunk 2 above)
Comment (mohamedgh16) [reply content/articles/polymorphism-in-java/index.md:209]: You could also replace the conclusion with the summary.

**Comment by Kimosop-Wakanda** [comment]
I have made the changes requested @mohamedgh16 
Please proceed to the tech review

File: content/articles/polymorphism-in-java/index.md [hunk 3]
```diff
@@ -45,14 +51,16 @@ Polymorphism can be divided into two:
...
+An example includes Method overloading or Function overloading.
 
 For clarity:
 
-`Method overloading` is achieved by having multiple methods to have the same name but different parameters. The parameters can differ by having different arguments, having different types of data types, or having different sizes the arguments. This will avoid the compile-time error.
+`Method overloading` is achieved by having multiple methods to have the same name but different parameters or having different sizes of arguments. The parameters can differ by having different arguments, having different types of data types, or having different sizes the arguments. This will avoid the compile-time error.
 
-Let us use a code snippet to try to understand more about static polymorphism.
+Let us use a code snippet understand more about static polymorphism.
```
Comment (mohamedgh16) [inline content/articles/polymorphism-in-java/index.md:63]: Let us use a code snippet to understand more about static polymorphism.

**Comment by mohamedgh16** [comment]
Tech review in progress @ahmadmardeni1 @Kimosop-Wakanda

**Comment by mohamedgh16** [comment]
Front matter
---
layout: engineering-education
status: publish
published: true
url: /polymorphism-in-java/
title: Polymorphism in Java
description: This article will help you understand Polymorphism in java.
author: Meshack Kimosop
date: 2021-1-19
topics: [Languages]
excerpt_separator: 
images:

  - url: /polymorphism-in-java/hero.jpg
    alt: Polymorphism in Java
---

Peer Review Contributions by: [Mohamed Alghadban](/engineering-education/authors/mohamed-alghadban/)

**Comment by mohamedgh16** [comment]
@Kimosop-Wakanda Edit the last comment.
@ahmadmardeni1 This article is ready for final review.

File: content/articles/polymorphism-in-java/index.md (hunk 3 above)
Comment (Kimosop-Wakanda) [reply content/articles/polymorphism-in-java/index.md]: Done

File: content/articles/polymorphism-in-java/index.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+Meow
+```
+
+In the example above, we have two classes, the dog class, and the cat class which both extend the class animal. You can note that when we called the object dog and cat, we receive the outputs for the methods in their respective classes.
+
+This indicates that the method in the class animal was overridden by the child class. Normally, for easier understanding of the code, we use `@Override` keyword to indicate which method is overriding the parent method. This gives a good example of dynamic polymorphism.
+
+### Advantages of Polymorphism
+We are now going to look at the advantages of each of the types of polymorphism.
+
```
Comment (ahmadmardeni1) [inline content/articles/polymorphism-in-java/index.md:174]: You have discussed the **Dynamic Polymorphism** advantages. But what about the **Static Polymorphism**. Or am I missing something?  @Kimosop-Wakanda 
cc @mohamedgh16

Comment (Kimosop-Wakanda) [reply content/articles/polymorphism-in-java/index.md]: I have made the correction @ahmadmardeni1

**Comment by Kimosop-Wakanda** [comment]
I have made the change @ahmadmardeni1 
cc: @mohamedgh16

**Comment by ahmadmardeni1** [comment]
Can you check if this info is correct please? @mohamedgh16

**Pull request merged**